docs/log-levels.md

```diff
@@ -22,17 +22,17 @@ In the following section we'll explore how to use them in practice.
 
 It is always fine for a library to log at `trace` and `debug` levels, and these two should be the primary levels any library is logging at.
 
-`trace` is the finest log level, and end-users of a library will not usually use it unless debugging very specific issues. You should consider it as a way for library developers to "log everything we could possibly need to diagnose a hard to reproduce bug." It is expected to take a toll on the performance of a system, and developers can assume trace level logging will not be used in production deployments, unless enabled specifically to locate some specific issue.
+`trace` is the finest log level, and end-users of a library will not usually use it unless debugging very specific issues. You should consider it as a way for library developers to "log everything we could possibly need to diagnose a hard to reproduce bug."  Unrestricted logging at `trace` level may take a toll on the performance of a system, and developers can assume trace level logging will not be used in production deployments, unless enabled specifically to locate some specific issue.
 
 This is in contrast with `debug` which some users _may_ choose to run enabled on their production systems.
 
 > Debug level logging should be not "too" noisy. Developers should assume some production deployments may need to (or want to) run with debug level logging enabled. 
 >
 > Debug level logging should not completely undermine the performance of a production system.
 
-As such, `debug` logging should not be overly noisy. It should provide a high value understanding of what is going on in the library for end users without diving deep into its internals. This is what `trace` is intended for.
+As such, `debug` logging should provide a high value understanding of what is going on in the library for end users, using domain relevant language.  Logging at `debug` level should not be overly noisy or dive deep into internals; this is what `trace` is intended for.
 
-Use `warning` level sparingly. Whenever possible, try to rather return or throw `Error` to end users that are descriptive enough so they can inspect, log them and figure out the issue. Potentially they may then enable debug logging to find out more about the issue.
+Use `warning` level sparingly. Whenever possible, try to rather return or throw `Error` to end users that are descriptive enough so they can inspect, log them and figure out the issue. Potentially, they may then enable debug logging to find out more about the issue.
```

Comment:
I generally think of warnings as when you can complete an operation, and errors when you can't.  This section describe returning or throwing errors, but throwing an `Error` implies that the operation itself fails which doesn't sound like a warning.

Reply:
It's like in Scala when you return a failed `Try` you wouldn't _also_ log that you're doing that, in order to avoid double-logging the same error. It's "throw or log" basically.

Reply:
I agree with "throw or log", it's attaching this advice specifically to `warning` (rather than any logging at all) that confuses it -- returning `scala.util.Failure` with a warning makes no sense, you'd log an error since the operation didn't complete.